Language/Haskell/GhcMod/Utils.hs

```diff
@@ -156,3 +160,41 @@ canonFilePath f = do
   e <- doesFileExist p
   when (not e) $ error $ "canonFilePath: not a file: " ++ p
   return p
+
+withMappedFile :: (IOish m, GmState m, GmEnv m) =>
+                  forall a. FilePath -> (FilePath -> m a) -> m a
+withMappedFile file action = getCanonicalFileNameSafe file >>= lookupMMappedFile >>= runWithFile
+  where
+    runWithFile (Just (RedirectedMapping to)) = action to
+    runWithFile (Just (MemoryMapping (Just src))) = do
+      crdl <- cradle
+      (fp,hndl) <- liftIO $ openTempFile (cradleTempDir crdl) (takeFileName file)
+      liftIO $ hPutStr hndl src
+      liftIO $ hClose hndl
+      result <- action fp
+      liftIO $ removeFile fp
+      return result
+    runWithFile _ = action file
+
+getCanonicalFileNameSafe :: (IOish m, GmEnv m) => FilePath -> m FilePath
+getCanonicalFileNameSafe fn = do
+  crdl <- cradle
+  let ccfn = cradleCurrentDir crdl </> fn
+  fex <- liftIO $ doesFileExist ccfn
+  if fex
+    then liftIO $ canonicalizePath ccfn
+    else return ccfn
+
+mkRevRedirMapFunc :: (Functor m, GmState m, GmEnv m) => m (FilePath -> FilePath)
```

Comment:
Long story short, it creates a "reverse redirection map", i.e. redirected file name as key and filename relative to `cradleRootDir` as value. Only reason it exists is to pass it to logging/reporting code, avoiding possible multiple O(n) lookups through "forward map" (that is returned by `getMMappedFiles`)